Add max bytes in line limit

aleksmaus opened the pull request:
## What does this PR do?

Handles OOM on long lines gracefully, writes the error into log.

- Addresses https://github.com/elastic/beats/issues/19500

## Why is it important?

Handle bad data gracefully without OOM.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ] I just recently started looking at the beats code so might not have enough knowledge of any possible side-effects of this change. Please review with that in mind.

## How to test this PR locally

Refer to https://github.com/elastic/beats/issues/19500 for test case.
The filebeat doesn't get killed due to OOM with this change and the error is logged (see the screenshot attached).

## Related issues

- Addresses https://github.com/elastic/beats/issues/19500

## Screenshots

<img width="1110" alt="Screen Shot 2020-07-01 at 9 16 06 AM" src="https://user-images.githubusercontent.com/872351/86249080-c5f92600-bb7c-11ea-9703-91d06e397fa8.png">

## Logs

```
2020-07-01T13:15:17.630Z	INFO	[crawler]	beater/crawler.go:108	Loading and starting Inputs completed. Enabled inputs: 1
2020-07-01T13:15:17.633Z	INFO	log/harvester.go:297	Harvester started for file: /test/readme.log
2020-07-01T13:15:17.635Z	ERROR	log/harvester.go:330	Read line error: exceeded max bytes in line limit; File: /test/readme.log
```

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

aleksmaus commented:
hmm, seems like a problem with CI

```
[2020-07-01T13:40:28.732Z] + mage package

[2020-07-01T13:40:28.732Z] /var/lib/jenkins/workspace/Beats_packaging_PR-19552/src/github.com/elastic/beats/auditbeat@tmp/durable-fb3d78f4/script.sh: 1: /var/lib/jenkins/workspace/Beats_packaging_PR-19552/src/github.com/elastic/beats/auditbeat@tmp/durable-fb3d78f4/script.sh: mage: not found

script returned exit code 127
```

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-19552/detail/PR-19552/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-19552/detail/PR-19552/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-19552/detail/PR-19552/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats-beats-mbp%2FPR-19552/detail/PR-19552/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_19552.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #19552 opened]`

* Start Time: 2020-07-01T13:34:40.133+0000

* Duration: 95 min 0 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 9614  |
| Skipped      | 1579 |
| Total        | 11193   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Report to Codecov`
    * Description:         curl -sSLo codecov https://codecov.io/bash        for i in auditbeat filebeat heartbeat libbeat metricbeat packetbeat winlogbeat journalbeat        do          FILE="${i}/build/coverage/full.cov"          if [ -f "${FILE}" ]; then            bash codecov -f "${FILE}"          fi        done      
    
    * Duration: 1 min 27 sec
    
    * Start Time: 2020-07-01T14:23:43.753+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats-beats-mbp/pipelines/PR-19552/runs/1/steps/3121/log)
    
  
  * **Name**: `Report to Codecov`
    * Description:         curl -sSLo codecov https://codecov.io/bash        for i in auditbeat filebeat heartbeat libbeat metricbeat packetbeat winlogbeat journalbeat        do          FILE="${i}/build/coverage/full.cov"          if [ -f "${FILE}" ]; then            bash codecov -f "${FILE}"          fi        done      
    
    * Duration: 1 min 27 sec
    
    * Start Time: 2020-07-01T14:19:48.631+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats-beats-mbp/pipelines/PR-19552/runs/1/steps/2932/log)
    
  
  * **Name**: `Report to Codecov`
    * Description:         curl -sSLo codecov https://codecov.io/bash        for i in auditbeat filebeat heartbeat libbeat metricbeat packetbeat winlogbeat journalbeat        do          FILE="${i}/build/coverage/full.cov"          if [ -f "${FILE}" ]; then            bash codecov -f "${FILE}"          fi        done      
    
    * Duration: 2 min 22 sec
    
    * Start Time: 2020-07-01T14:10:53.772+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats-beats-mbp/pipelines/PR-19552/runs/1/steps/2363/log)
    
  
  * **Name**: `Report to Codecov`
    * Description:         curl -sSLo codecov https://codecov.io/bash        for i in auditbeat filebeat heartbeat libbeat metricbeat packetbeat winlogbeat journalbeat        do          FILE="${i}/build/coverage/full.cov"          if [ -f "${FILE}" ]; then            bash codecov -f "${FILE}"          fi        done      
    
    * Duration: 2 min 22 sec
    
    * Start Time: 2020-07-01T14:13:15.868+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats-beats-mbp/pipelines/PR-19552/runs/1/steps/2433/log)
    
  
  </p>
  </details>




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

aleksmaus commented:
The today's comment on the original issue was a suggestion to skip the long lines instead of returning the error. Will update PR later today.

aleksmaus commented:
Updated PR to skip the long lines, logging the warning and continuing processing.

New log screenshot, running on the 100M suggested in the test case with couple of lines appended.
Shows skipping the long line. Added more unit test coverage for skipping the long lines.

Using the same MaxSize from the config that is used for the limiter at the end of the readers pipeline, multiplied by 4 since the checks and skipping happen before decoding the data.

![Screen Shot 2020-07-02 at 11 57 32 AM](https://user-images.githubusercontent.com/872351/86382322-bb618e00-bc5b-11ea-8e2e-8a039801124f.png)

benbuzbee commented on filebeat/input/log/harvester.go:
```diff
@@ -644,10 +646,17 @@ func (h *Harvester) newLogFileReader() (reader.Reader, error) {
+	// for the worst case scenario where incoming UTF32 charchers are decoded to the single byte UTF-8 characters.
+	// This limit serves primarily to avoid memory bload or potential OOM with expectedly long lines in the file.
+	// The further size limiting is performed by LimitReader at the end of the readers pipeline as needed.
+	encReaderMaxBytes := h.config.MaxBytes * 4
```
may want to add a comment in docs for max_bytes that it can use up to 4x the number worth of memory for buffering

aleksmaus replied:
Sure can do that. Which docs? Before this change it was taking all available memory :-)

benbuzbee replied:
https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-log.html

As a user I would want that information here


The maximum number of bytes that a single log message can have. All bytes after max_bytes are discarded and not sent., _but up to 4x max_bytes will be buffered in-memory_ .This setting is especially useful for multiline log messages, which can get large. The default is 10MB (10485760).

aleksmaus replied:
On the second thought I'm not sure that this part of internal implementation should be documented. 
Before this change the code was allocating as much as it was available and cutting the final message to the max byte limit before the output from the pipeline.
This change adds a protection from unlimited allocation.

I'll let the regular beats contributors that are going to review this PR to decide if this needs to be documented.

benbuzbee replied:
Fair enough. I will make a quick pitch and leave it to you and reviewers to decide: without that kind of information, it is hard to plan. For example, how can I determine how much memory I will be safe giving filebeat? I may think that if I tell it to read only 1 file at a time, and tell it to use max_bytes of 25MB, that I am very safe giving it 50MB of memory, only to be sad when it OOMs because it actually used 100MB of memory for 1 line that I thought was limited to 25.

Personally I find it already very hard to figure out how much memory it needs because i have to do match like <overhead> + max_events * num_harvesters * max_size - and im scared of more hidden values (like * 4)

aleksmaus commented on libbeat/reader/readfile/line.go:
```diff
@@ -147,6 +154,11 @@ func (r *LineReader) advance() error {
+
+		// If newLine is not found and exceeded max bytes limit
+		if idx == -1 && r.maxBytes != 0 && r.inBuffer.Len() > r.maxBytes {
+			return ErrExceededMaxBytesLimit
```
I think you commented on the first commit. This code was changed by the second commit in this PR.

aleksmaus commented:
Can reduce the memory footprint by not caching the line data before decoding and decoding every read. Need to measure how it affects the performance. 
This would require a bit more changes in the exiting code.
The current change is less invasive. 
Will leave up to the reviewers feedback.

aleksmaus commented:
@kvch if I'm reading the implementation correctly https://github.com/elastic/beats/blob/efa5ebade3c7eae3af5f44cbd27c4207671a2bb6/libbeat/common/streambuf/streambuf.go#L129

it caps the the number of byte in append operation, while potentially doing huge (10MB default) allocations for backing array: 
https://github.com/elastic/beats/blob/efa5ebade3c7eae3af5f44cbd27c4207671a2bb6/libbeat/common/streambuf/streambuf.go#L145

and not really limiting the buffer size growth
https://github.com/elastic/beats/blob/efa5ebade3c7eae3af5f44cbd27c4207671a2bb6/libbeat/common/streambuf/streambuf.go#L158

Please explain, if I misunderstood the way you are proposing to use this?

kvch commented on libbeat/reader/readfile/line.go:
```diff
@@ -28,12 +28,15 @@ import (
 	"github.com/elastic/beats/v7/libbeat/logp"
 )
 
+const Unlimited = 0
```
Please do not export this variable if you are not using it outside the `readfile` package.

aleksmaus replied:
This constant potentially could be used for the public NewLineReader config param.

kvch replied:
Personally, I am not a huge fan of exposing something we would potentially use. In general, it is easier to export something unexported, then unexport something exported.

But I do not block this PR because of a philosophical question about a small constant.

aleksmaus replied:
Sounds good, will update to not export.

kvch commented on libbeat/reader/readfile/line_test.go:
```diff
@@ -31,10 +35,12 @@ import (
 
 // Sample texts are from http://www.columbia.edu/~kermit/utf8.html
-var tests = []struct {
+type LineTestSpec struct {
```
Please do not export this struct as you are not using it from outside the package. Also, rather name it `listTestCase`. The name you have given it is not idiomatic to this repo.

aleksmaus replied:
Updated

kvch commented on libbeat/reader/readfile/line_test.go:
```diff
@@ -248,3 +255,129 @@ func testReadLines(t *testing.T, inputLines [][]byte) {
+}
+
+func TestMaxBytesLimit(t *testing.T) {
+	const (
```
Could you please extract the setup part of the test out into a separate function? The way the test is structured now is hard to read and understand.

aleksmaus replied:
Extracted.

kvch commented:
@aleksmaus You are right. I was just wondering if we could use it.

aleksmaus commented:
Addressed all of the comments. Let me know if anything else needs to be changed.

kvch commented:
jenkins run tests

kvch commented:
@aleksmaus Sorry for the long wait. The code in general looks good to me. However, when a line is truncated Filebeat flags that message by adding "truncated" to the list in `log.flags`. With your current implementation, it is not possible to propagate this information to the appropriate readers. Could you please address this problem?

aleksmaus commented:
@kvch Sure, no worries. Didn't know about this flag. Will check what can be done.

aleksmaus commented:
@kvch I checked the existing code it looks like the "truncated" flag is only added to the lines that are truncated in the LimitReader at the end of pipeline after decoding. 
In this PR we completely drop the line that is too long (logging it) to avoid OOM, 
so there is no "message" and there is nothing to attach that flag to.
We can't safely truncate the line bytes before decoding them either.
Thoughts?

kvch commented:
@aleksmaus Indeed. Could you please rebase the PR and add a changelog entry to `CHANGELOG.next.asciidoc`?

aleksmaus commented:
> > we completely drop the line that is too long (logging it)
> 
> Have we also considered truncating the line before logging? We have seen ~1Gb of junk causing filebeat to OOM. It might be helpful to truncate first before logging. Perhaps only log the first 1024 characters or something for the dropped line?

the line drop happens before decoding. would have to try to decode for that. the number of characters will be known only after decoding. also the junk will not be necessarily decodable.
thoughts?

aleksmaus commented:
> > we completely drop the line that is too long (logging it)
> 
> Have we also considered truncating the line before logging? We have seen ~1Gb of junk causing filebeat to OOM. It might be helpful to truncate first before logging. Perhaps only log the first 1024 characters or something for the dropped line?

Ahh, not logging the line content itself. Just the fact that the line is dropped and the length in bytes.

kvch commented:
@aleksmaus Thank you for the contribution!
